### DISCUSSION TITLE
What causes a `400` error with message `Server not found for <url>`?

### DISCUSSION THREAD

**Discussion by jtherrmann** [p1c2u/openapi-core#757] [Q&A] [answered]
All of the API endpoints for the test deployment of our Flask app, which is deployed via Amazon API Gateway, are currently returning a response like the one below:

```
{
  "detail": "Server not found for https://vd2gh6uqw3.execute-api.us-west-2.amazonaws.com/api/user",
  "status": 400,
  "title": "Bad Request",
  "type": "about:blank"
}
```

We depend on the latest pre-release of `openapi-core`, <https://github.com/python-openapi/openapi-core/releases/tag/0.19.0a1>.

The error message leads me to believe that `openapi-core` is raising a [`ServerNotFound`](https://github.com/python-openapi/openapi-core/blob/0.19.0a1/openapi_core/templating/paths/exceptions.py#L40) exception from `BasePathFinder.find` ([here](https://github.com/python-openapi/openapi-core/blob/0.19.0a1/openapi_core/templating/paths/finders.py#L51)), which is called from `BaseAPICallValidator._find_path` ([here](https://github.com/python-openapi/openapi-core/blob/0.19.0a1/openapi_core/validation/validators.py#L277)).

I have a couple of questions:
* Is `BaseAPICallValidator` for validating requests, responses, or both?
* Why does the validator need to validate the server URL? What exactly does [`APICallPathFinder._get_servers_iter`](https://github.com/python-openapi/openapi-core/blob/0.19.0a1/openapi_core/templating/paths/finders.py#L97) do? What is the fundamental meaning of a `ServerNotFound` exception?

**Comment by p1c2u** [comment, marked as answer]
HI @jtherrmann 

the error "Server not found for https://vd2gh6uqw3.execute-api.us-west-2.amazonaws.com/api/user" means your OpenAPI specification doesn't define server that match url from your request.

> Is BaseAPICallValidator for validating requests, responses, or both?

It's a base class for request and response validation

> Why does the validator need to validate the server URL? 

Server is part of OpneAPI specification

> What exactly does [APICallPathFinder._get_servers_iter](https://github.com/python-openapi/openapi-core/blob/0.19.0a1/openapi_core/templating/paths/finders.py#L97) do? 

Finds valid server definition in your specification

> What is the fundamental meaning of a ServerNotFound exception?

`ServerNotFound` exception is raised when your request server does not match any defined servers in your specification.

**Reply by jtherrmann** [reply]
@p1c2u This answer was extremely helpful, thank you! I had posted some comments with additional questions, but I deleted them because they were based on a partial misunderstanding of how OpenAPI works. But I understand how server URLs work now and we were able to resolve the issues with our app. Thanks again for your quick response!

**Reply by jtherrmann** [reply]
In case anyone reading this is having a similar issue, our problem was that we have an AWS CloudFront distribution that forwards API requests to our AWS ApiGateway app. The forwarded requests have a URL like `<api-gateway-url>/api/<endpoint>`, e.g. `https://vd2gh6uqw3.execute-api.us-west-2.amazonaws.com/api/user`, but our app does not expect the `/api` component. Fixing this issue is as simple as specifying the following relative server URLs in our OpenAPI spec:

```
servers:
  - url: /
  - url: /api
```

The `/` url is technically optional, but including it allows us to make queries via our Swagger UI (which would force us to send requests to `/api` if that were the only server url specified).

**Reply by p1c2u** [reply]
I'm glad you found the problem. Any suggestions about debugging improvements are welcome.

**Reply by jtherrmann** [reply]
@p1c2u I think the error message for the `ServerNotFound` exception could be improved. When I first saw the error:

`Server not found for https://vd2gh6uqw3.execute-api.us-west-2.amazonaws.com/api/user`

I assumed that the request had failed to reach the server. I did not even realize that the error was coming from the OpenAPI spec validator. At a minimum, I think the error message should make it clear that this is an OpenAPI spec validation problem. For example, it could say something like this:

`Request is not valid for the given OpenAPI spec: Request URL https://vd2gh6uqw3.execute-api.us-west-2.amazonaws.com/api/user does not match any of the OpenAPI spec server URLs`

In fact, I think that all request validation errors should begin with something like `Request is not valid for the given OpenAPI spec:` to make it clear that this is an OpenAPI spec error.
